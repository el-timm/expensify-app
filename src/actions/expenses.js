import uuid from 'uuid';
import database from '../firebase/firebase';

// React + Redux (only)
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// React + Firebase (+ Redux)
// component calls action generator
// action generator returns fuction
// component dispatches fnction (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)
//    returns an object, which manipulates the Redux store


// ADD_EXPENSE
export const addExpense = (expense) => ({
        type: 'ADD_EXPENSE',
        expense
});

// uses redux thunk
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});