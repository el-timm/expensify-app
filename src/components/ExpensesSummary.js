import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => (
    <div>
        <h1>
            Viewing {expenseCount}
            {expenseCount === 1 ? ' expense ' : ' expenses ' }
            totalling {numeral(expenseTotal / 100).format('$0,0.00')} 
        </h1>
    </div>
);

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    console.log(visibleExpenses);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: getTotalExpenses(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);