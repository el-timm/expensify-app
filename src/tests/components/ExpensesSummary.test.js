import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('render ExpensesSummary (no expenses)', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={0} expenseTotal={0} />);
    expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary (1 expense)', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={12300} />);
    expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary (3 expenses)', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expenseTotal={1234500} />);
    expect(wrapper).toMatchSnapshot();
});