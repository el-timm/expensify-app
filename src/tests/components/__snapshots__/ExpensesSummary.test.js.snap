// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpensesSummary (1 expense) 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense 
    totalling 
    $123.00
  </h1>
</div>
`;

exports[`render ExpensesSummary (3 expenses) 1`] = `
<div>
  <h1>
    Viewing 
    3
     expenses 
    totalling 
    $12,345.00
  </h1>
</div>
`;

exports[`render ExpensesSummary (no expenses) 1`] = `
<div>
  <h1>
    Viewing 
    0
     expenses 
    totalling 
    $0.00
  </h1>
</div>
`;
